modules/space/kepler.cpp

```diff
@@ -0,0 +1,637 @@
+        const double epoch = nSecondsSince2000 + nLeapSecondsOffset - nSecondsEpochOffset;
+        return epoch;
+    }
+
+    double epochFromYMDdSubstring(const std::string& epochString) {
+        // The epochString is in the form:
+        // YYYYMMDD.ddddddd
+        // With YYYY as the year, MM the month (1 - 12), DD the day of month (1-31),
+        // and dddd the fraction of that day.
+
+        // The main overview of this function:
+        // 1. Read the year value
+        // 2. Calculate the number of days since the beginning of the year
+        // 3. Convert the number of days to a number of seconds
+        // 4. Get the number of leap seconds since January 1st, 2000 and remove them
+        // 5. Adjust for the fact the epoch starts on 1st January at 12:00:00, not
+        // midnight
+
+        // 1
+        int year = std::atoi(epochString.substr(0, 4).c_str());
+        const int daysSince2000 = countDays(year);
+
+        // 2.
+        int monthNum = std::atoi(epochString.substr(4, 2).c_str());
+        int dayOfMonthNum = std::atoi(epochString.substr(6, 2).c_str());
+        int wholeDaysInto = daysIntoGivenYear(year, monthNum, dayOfMonthNum);
+        double fractionOfDay = std::atof(epochString.substr(9, 7).c_str());
+        double daysInYear = static_cast<double>(wholeDaysInto) + fractionOfDay;
+
+        // 3
+        using namespace std::chrono;
+        const int SecondsPerDay = static_cast<int>(seconds(hours(24)).count());
+        //Need to subtract 1 from daysInYear since it is not a zero-based count
+        const double nSecondsSince2000 = (daysSince2000 + daysInYear - 1) * SecondsPerDay;
+
+        // 4
+        // We need to remove additional leap seconds past 2000 and add them prior to
+        // 2000 to sync up the time zones
+        const double nLeapSecondsOffset = -countLeapSeconds(
+            year,
+            static_cast<int>(std::floor(daysInYear))
+        );
+
+        // 5
+        const double offset = static_cast<double>(seconds(hours(12)).count());
+
+        // Combine all of the values
+        const double epoch = nSecondsSince2000 + nLeapSecondsOffset - offset;
+        return epoch;
+    }
+
+    double epochFromOmmString(const std::string& epochString) {
+        // The epochString is in the form:
+        // YYYY-MM-DDThh:mm:ss[.d->d][Z]
+        // or
+        // YYYY-DDDThh:mm:ss[.d->d][Z]
+
+        // The main overview of this function:
+        // 0. Determine which type it is
+        // 1. Read the year value
+        // 2. Calculate the number of days since the beginning of the year
+        // 3. Convert the number of days to a number of seconds
+        // 4. Get the number of leap seconds since January 1st, 2000 and remove them
+        // 5. Add the hh:mm:ss component
+        // 6. Adjust for the fact the epoch starts on 1st January at 12:00:00, not
+        // midnight
+
+        // 1
+        int year = std::atoi(epochString.substr(0, 4).c_str());
```

Comment:
There should probably be some format checking of `epochString` before using `std::atoi`, since it can return a zero value (or undefined behavior possibly meaning any int value) if the string contains unexpected characters. The check would verify the YYYY-MM-DD.... expected form.
This comment applies to other uses of `std::atoi` here as well.

Reply:
Going to change that in the next PR.  I'm flirting with https://github.com/eliaskosunen/scnlib right now and would like to use that instead of manually parsing and checking for errors